README.md
# @bpmn-io/align-to-origin

[![CI](https://github.com/bpmn-io/align-to-origin/workflows/CI/badge.svg)](https://github.com/bpmn-io/align-to-origin/actions?query=workflow%3ACI)


Aligns your diagrams to the coordinate origin `(0,0)`, manually triggered or on diagram save.

Plugs into your favorite [BPMN](https://github.com/bpmn-io/bpmn-js), [DMN](https://github.com/bpmn-io/dmn-js) and [CMMN](https://github.com/bpmn-io/cmmn-js) editor.

<img src="https://raw.githubusercontent.com/bpmn-io/align-to-origin/master/resources/screencapture.gif" width="600" alt="Automatic Origin Adjustment" />

_Alignment in action. [Diagram origin marker](https://github.com/bpmn-io/diagram-js-origin), alignment corridor, and element boxes shown for demonstration purposes only._


## Usage

```javascript
import BpmnModeler from 'bpmn-js/lib/Modeler';

import AlignToOrigin from '@bpmn-io/align-to-origin';


// extend the BPMN editor with the exporter module
const modeler = new BpmnModeler({
  alignToOrigin: {
    alignOnSave: true,
    offset: 150,
    tolerance: 50
  },
  additionalModules: [
    AlignToOrigin
  ]
});


// hooks into #saveXML to align the diagram elements
modeler.saveXML(function(err, xml) {
  ...
});

// may be used standalone, too
const alignToOrigin = modeler.get('alignToOrigin');

alignToOrigin.align();
```


## License

MIT


CHANGELOG.md
# Changelog

All notable changes to [@bpmn-io/align-to-origin](https://github.com/bpmn-io/align-to-origin) are documented here. We use [semantic versioning](http://semver.org/) for releases.

## Unreleased

___Note:__ Yet to be released changes appear here._

## 0.7.0

* `CHORE`: migrate CI to GitHub Actions ([`a3c349de`](https://github.com/bpmn-io/align-to-origin/commit/a3c349de675fa514043fef7f1cb46e230cac589b))
* `CHORE`: add autosaving integration test ([#5](https://github.com/bpmn-io/align-to-origin/pull/5))
* `CHORE`: early abort alignment if no elements ([`d6c5580d`](https://github.com/bpmn-io/align-to-origin/commit/d6c5580da8cbf3db4b455ec9ceab41221534b845))

## 0.6.0

* `FEAT`: make grid aware

## 0.5.0

* `CHORE`: update `y` offset to more sane default

## 0.4.2

* `FIX`: move canvas after `commandStack.changed` ([`95834a8439fc3f053995740d8dc8ce7302b3be39`](https://github.com/bpmn-io/align-to-origin/commit/95834a8439fc3f053995740d8dc8ce7302b3be39))

## 0.4.1

* `FIX`: correct documentation

## 0.4.0

* `FEAT`: always compensate for element movement by scrolling canvas, drop `scrollCanvas` option
* `FEAT`: implement canvas scrolling in an undoable manner

## 0.3.0

* `FEAT`: add ability to specify `{ x, y }` offset
* `FEAT`: adjust `x` and `y` independent from each other
* `FEAT`: apply sane default for `offset` and `tolerance`
* `CHORE`: rename `padding` to `offset`
* `CHORE`: rename `thresold` to `tolerance`

## 0.2.0

* `FEAT`: make `alignOnSave=true` the default behavior

## 0.1.0

_Initial version._

test/test.js
import BpmnModeler from 'bpmn-js/lib/Modeler';
import DmnModeler from 'dmn-js/lib/Modeler';
import CmmnModeler from 'cmmn-js/lib/Modeler';

import { insertCSS } from 'bpmn-js/test/helper';

import TestVisualizationModule from './test-visualization';
import DiagramOriginModule from 'diagram-js-origin';

import AlignToOriginModule from '..';


var NoGridSnapping = {
  'gridSnapping': [ 'value', null ]
};

insertCSS(
  'test.css',
  require('./test.css')
);

insertCSS(
  'diagram.css',
  require('bpmn-js/dist/assets/diagram-js.css')
);

insertCSS(
  'bpmn-font.css',
  require('bpmn-js/dist/assets/bpmn-font/css/bpmn-embedded.css')
);


describe('alignToOrigin', function() {

  it.skip('should work (visual test)', function(done) {

    // given
    var diagramXML = require('./process.bpmn');

    var modeler = new BpmnModeler({
      container: 'body',
      additionalModules: [
        TestVisualizationModule,
        DiagramOriginModule,
        AlignToOriginModule
      ],
      keyboard: {
        bindTo: document.body
      }
    });

    setInterval(function() {
      modeler.get('alignToOrigin').align();
    }, 3000);

    modeler.importXML(diagramXML, done);

  });


  it('should extend BpmnModeler instance without grid snapping', function(done) {

    // given
    var diagramXML = require('./process.bpmn');

    var modeler = new BpmnModeler({
      container: 'body',
      additionalModules: [
        AlignToOriginModule,
        NoGridSnapping
      ]
    });

    var elementRegistry = modeler.get('elementRegistry');

    modeler.importXML(diagramXML, function() {

      var element = elementRegistry.get('StartEvent');

      // when
      modeler.saveXML(function(err, xml) {

        // then
        // expect element got aligned
        expect(element.x).to.eql(156);
        expect(element.y).to.eql(81);

        done(err);
      });

    });

  });


  it('should extend BpmnModeler instance with grid snapping', function(done) {

    // given
    var diagramXML = require('./process.bpmn');

    var modeler = new BpmnModeler({
      container: 'body',
      additionalModules: [
        AlignToOriginModule
      ]
    });

    var elementRegistry = modeler.get('elementRegistry');

    modeler.importXML(diagramXML, function() {

      var element = elementRegistry.get('StartEvent');

      // when
      modeler.saveXML(function(err, xml) {

        // then
        // expect element got aligned
        expect(element.x).to.eql(156);
        expect(element.y).to.eql(84);

        done(err);
      });

    });

  });


  it('should disable alignOnSave', function(done) {

    // given
    var diagramXML = require('./process.bpmn');

    var modeler = new BpmnModeler({
      container: 'body',
      additionalModules: [ AlignToOriginModule ],
      alignToOrigin: {
        alignOnSave: false
      }
    });

    var elementRegistry = modeler.get('elementRegistry');

    modeler.importXML(diagramXML, function() {

      var element = elementRegistry.get('StartEvent');

      // when
      modeler.saveXML(function(err, xml) {

        // then
        // expect element got not aligned
        expect(element.x).not.to.eql(156);
        expect(element.y).not.to.eql(156);

        done(err);
      });
    });

  });


  it('should extend CmmnModeler instance', function(done) {

    // given
    var diagramXML = require('./case.cmmn');

    var modeler = new CmmnModeler({
      container: 'body',
      additionalModules: [ AlignToOriginModule ]
    });

    var elementRegistry = modeler.get('elementRegistry');

    modeler.importXML(diagramXML, function() {

      var element = elementRegistry.get('CasePlanModel');

      // when
      modeler.saveXML(function(err, xml) {

        // then
        // expect element got aligned
        expect(element.x).to.eql(156);
        expect(element.y).to.eql(99);

        done(err);
      });
    });

  });


  it('should extend Drd instance', function(done) {

    // given
    var diagramXML = require('./decision.dmn');

    var modeler = new DmnModeler({
      container: 'body',
      drd: {
        additionalModules: [ AlignToOriginModule ]
      }
    });

    modeler.importXML(diagramXML, function() {

      var elementRegistry = modeler.getActiveViewer().get('elementRegistry');

      var element = elementRegistry.get('Decision');

      // when
      modeler.saveXML(function(err, xml) {

        // then
        // expect element got aligned
        expect(element.x).to.eql(200);
        expect(element.y).to.eql(81);

        done(err);
      });
    });

  });


  it('should compute correct adjustment', function() {

    // given
    var modeler = new BpmnModeler({
      container: 'body',
      additionalModules: [ AlignToOriginModule ]
    });

    var alignToOrigin = modeler.get('alignToOrigin');

    var config = {
      offset: { x: 100, y: 100 },
      tolerance: 50
    };

    // then
    expect(alignToOrigin.computeAdjustment({ x: 40, y: 60 }, config)).to.eql({ x: 60, y: 0 });
    expect(alignToOrigin.computeAdjustment({ x: 60, y: 40 }, config)).to.eql({ x: 0, y: 60 });
    expect(alignToOrigin.computeAdjustment({ x: 140, y: 160 }, config)).to.eql({ x: 0, y: -60 });
    expect(alignToOrigin.computeAdjustment({ x: 160, y: 140 }, config)).to.eql({ x: -60, y: 0 });
    expect(alignToOrigin.computeAdjustment({ x: 90, y: 90 }, config)).to.eql({ x: 0, y: 0 });
    expect(alignToOrigin.computeAdjustment({ x: -30, y: 200 }, config)).to.eql({ x: 130, y: -100 });
  });


  it('should scroll canvas', function(done) {

    // given
    var diagramXML = require('./process.bpmn');

    var modeler = new BpmnModeler({
      container: 'body',
      additionalModules: [ AlignToOriginModule ],
      canvas: {
        deferUpdate: false
      }
    });


    var alignToOrigin = modeler.get('alignToOrigin');
    var elementRegistry = modeler.get('elementRegistry');
    var modeling = modeler.get('modeling');
    var commandStack = modeler.get('commandStack');
    var canvas = modeler.get('canvas');

    function expectAligned() {
      var inner = canvas.viewbox().inner;

      expect({ x: inner.x, y: inner.y }).to.eql({ x: 150, y: 78 });
    }

    modeler.importXML(diagramXML, function() {

      // given
      var element = elementRegistry.get('StartEvent');

      alignToOrigin.align();

      // assume
      expectAligned();

      // when
      modeling.moveElements([ element ], { x: -200, y: -200 });
      alignToOrigin.align();

      // then
      expectAligned();

      // when zooming in...
      canvas.zoom(1.3);

      // and undoing...
      commandStack.undo();
      commandStack.undo();

      // then
      expectAligned();

      done();
    });

  });

});

test/spec/auto-save/index.js
var LOW_PRIORITY = 250;

/**
 * Saves the diagram, on change.
 *
 * @constructor
 *
 * @param {didi.Injector} injector
 * @param {Object} config save config
 *
 * @example
 *
 * ```javascript
 * // once used, components may safely hook into `saveXML.done`
 * // to get notified on every save
 *
 * const modeler = new BpmnModeler({
 *   additionalModules: [
 *     AutoSaveModule
 *   ]
 * });
 *
 * modeler.on('saveXML.done', function(event) {
 *   const {
 *     error,
 *     xml
 *   } = event;
 *
 *   // ...
 * });
 * ```
 */
function AutoSave(injector, config) {

  config = config || { format: true };

  var parent = injector.get('_parent', false);

  var modeler = parent || injector.get('bpmnjs', false) || injector.get('cmmnjs');

  if (!modeler) {
    throw new Error('no modeler context');
  }

  var eventBus = parent && parent._eventBus || injector.get('eventBus');

  function subscribeChanged() {
    eventBus.once('commandStack.changed', LOW_PRIORITY, function() {
      modeler.saveXML(config)
        .catch(function(error) {
          console.error('[auto-save] failed', error);
        })
        .finally(subscribeChanged);
    });
  }

  subscribeChanged();
}

AutoSave.$inject = [ 'injector', 'config.autoSave' ];


export default {
  __init__: [
    'autoSave'
  ],
  autoSave: [ 'type', AutoSave ]
};

karma.conf.js
'use strict';

// use puppeteer provided Chrome for testing
process.env.CHROME_BIN = require('puppeteer').executablePath();

// configures browsers to run test against
// any of [ 'ChromeHeadless', 'Chrome', 'Firefox', 'IE', 'PhantomJS' ]
var browsers = (process.env.TEST_BROWSERS || 'ChromeHeadless').split(',');


module.exports = function(karma) {
  karma.set({

    frameworks: [
      'mocha',
      'chai',
      'webpack'
    ],

    files: [
      'test/testBundle.js'
    ],

    preprocessors: {
      'test/testBundle.js': [ 'webpack' ]
    },

    reporters: 'dots',

    browsers,

    browserNoActivityTimeout: 30000,

    autoWatch: false,
    singleRun: true,

    webpack: {
      mode: 'development',
      module: {
        rules: [
          {
            test: /\.(bpmn|dmn|cmmn|css)$/,
            use: 'raw-loader'
          }
        ]
      }
    }
  });

};


lib/align-to-origin.js
import {
  assign,
  isNumber
} from 'min-dash';

var DEFAULT_OPTIONS = {
  offset: {
    x: 150,
    y: 75
  },
  tolerance: 50,
  alignOnSave: true
};

var HIGHER_PRIORITY = 1250;


/**
 * Moves diagram contents to the origin + offset,
 * optionally upon diagram save.
 *
 * @param {Object} config
 * @param {didi.Injector} injector
 * @param {EventBus} eventBus
 * @param {CommandStack} commandStack
 * @param {Canvas} canvas
 * @param {Modeling} modeling
 */
export default function AlignToOrigin(config, injector, eventBus, commandStack, canvas, modeling) {

  /**
   * Return actual config with defaults applied.
   */
  function applyDefaults(config) {

    var c = assign({}, DEFAULT_OPTIONS, config || {});

    if (isNumber(c.offset)) {
      c.offset = {
        x: c.offset,
        y: c.offset
      };
    }

    return c;
  }

  config = applyDefaults(config);

  /**
   * Compute adjustment given the specified diagram origin.
   *
   * @param {Point} origin
   *
   * @return {Point} adjustment
   */
  function computeAdjustment(origin, config) {

    var offset = config.offset,
        tolerance = config.tolerance;

    var adjustment = {};

    [ 'x', 'y' ].forEach(function(axis) {

      var delta = -origin[axis] + offset[axis];

      var gridSnapping = injector.get('gridSnapping', false);

      if (gridSnapping) {
        delta = quantize(delta, gridSnapping.getGridSpacing());
      }

      adjustment[axis] = Math.abs(delta) < tolerance ? 0 : delta;
    });

    return adjustment;
  }


  /**
   * Align the diagram content to the origin.
   *
   * @param {Object} options
   */
  function align() {

    var bounds = canvas.viewbox().inner;

    var elements = canvas.getRootElement().children;

    if (!elements.length) {
      return;
    }

    var delta = computeAdjustment(bounds, config);

    if (delta.x === 0 && delta.y === 0) {
      return;
    }

    commandStack.execute('elements.alignToOrigin', {
      elements: elements,
      delta: delta
    });
  }


  /**
   * Setup align on save functionality
   */
  function bindOnSave() {

    // nested editors expose _parent to access the
    // save responsible entity
    var parent = injector.get('_parent', false);

    var localEvents = eventBus;

    var parentEvents = parent && parent._eventBus;

    (parentEvents || localEvents).on('saveXML.start', HIGHER_PRIORITY, align);

    if (parentEvents) {

      // unregister for saveXML.start
      localEvents.on('diagram.destroy', function() {
        parentEvents.off('saveXML.start', align);
      });
    }
  }


  /**
   * Create a function that compensates the element movement
   * by moving applying the delta in the given direction.
   */
  function movementCompensator(direction) {

    /**
     * Handler to executed
     */
    return function(context) {

      // adjust canvas after the commandstack got changed
      eventBus.once('commandStack.changed', function() {

        var delta = context.delta;
        var scale = canvas.viewbox().scale;

        canvas.scroll({
          dx: direction * delta.x * scale,
          dy: direction * delta.y * scale
        });
      });
    };
  }

  // command registration

  /**
   * A command handler that compensates the element movement
   * by applying the inverse move operation on the canvas.
   */
  commandStack.register('elements.alignToOrigin', {

    preExecute: function(context) {
      var delta = context.delta,
          elements = context.elements;

      modeling.moveElements(elements, delta);
    },

    execute: movementCompensator(-1),
    revert: movementCompensator(1)
  });

  // setup

  if (config.alignOnSave) {
    bindOnSave();
  }

  // API

  this.align = align;
  this.computeAdjustment = computeAdjustment;

  // internal debugging purposes
  this._config = config;
}

AlignToOrigin.$inject = [
  'config.alignToOrigin',
  'injector',
  'eventBus',
  'commandStack',
  'canvas',
  'modeling'
];


// helpers /////////////////////////

function quantize(value, quantum, fn) {
  if (!fn) {
    fn = 'round';
  }

  return Math[ fn ](value / quantum) * quantum;
}